integrations/git_annex_test.go

```diff
@@ -0,0 +1,940 @@
+   find the path in .git/annex/objects/ that contains its actual contents. */
+func annexObjectPath(repoPath string, file string) (string, error) {
+
+	// find the path inside *.git/annex/objects of a given file
+	// i.e. figure out its two-level hash prefix: https://git-annex.branchable.com/internals/hashing/
+	// ASSUMES the target file is checked into HEAD
+
+	var bare bool // whether the repo is bare or not; this changes what the hashing algorithm is, due to backwards compatibility
+
+	bareStr, _, err := git.NewCommand(git.DefaultContext, "config", "core.bare").RunStdString(&git.RunOpts{Dir: repoPath})
+	if err != nil {
+		return "", err
+	}
+
+	if bareStr == "true\n" {
+		bare = true
+	} else if bareStr == "false\n" {
+		bare = false
+	} else {
+		return "", errors.New(fmt.Sprintf("Could not determine if %s is a bare repo or not; git config core.bare = <%s>", repoPath, bareStr))
+	}
+
+	// given a repo and a file in it
+	// TODO: handle other branches, e.g. non-HEAD branches etc
+	annexKey, _, err := git.NewCommand(git.DefaultContext, "show", "HEAD:"+file).RunStdString(&git.RunOpts{Dir: repoPath})
+	if err != nil {
+		return "", err
+	}
+
+	annexKey = strings.TrimSpace(annexKey)
+	if !strings.HasPrefix(annexKey, "/annex/objects/") {
+		return "", errors.New(fmt.Sprintf("%s/%s does not appear to be annexed .", repoPath, file))
+	}
+	annexKey = strings.TrimPrefix(annexKey, "/annex/objects/")
+
+	var keyformat string
+	if bare {
+		keyformat = "hashdirlower"
+	} else {
+		keyformat = "hashdirmixed"
+	}
+	keyHashPrefix, _, err := git.NewCommand(git.DefaultContext, "annex", "examinekey", "--format=${"+keyformat+"}", annexKey).RunStdString(&git.RunOpts{Dir: repoPath})
+	if err != nil {
+		return "", err
+	}
+
+	if !bare {
+		repoPath = path.Join(repoPath, ".git")
+	}
+	return path.Join(repoPath, "annex", "objects", keyHashPrefix, annexKey, annexKey), nil
+}
+
+/*
+Do chmod -R +w $REPOS in order to handle https://git-annex.branchable.com/internals/lockdown/:
+
+> (The only bad consequence of this is that rm -rf .git doesn't work unless you first run chmod -R +w .git)
+
+Without, these tests can only be run once, because they reuse `gitea-repositories/`
+folder and will balk at finding pre-existing partial repos.
+*/
+func annexUnlockdown() {
+	filepath.WalkDir(setting.RepoRootPath, func(path string, d fs.DirEntry, err error) error {
+		if err == nil {
+			// 0200 == u+w, in octal unix permission notation
+			info, err := d.Info()
+			if err != nil {
+				return err
+			}
+
+			err = os.Chmod(path, info.Mode()|0200)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+/* like withKeyFile(), but automatically sets it the account given in ctx for use by git-annex */
+func withAnnexCtxKeyFile(t *testing.T, ctx APITestContext, callback func()) {
+	os.Setenv("GIT_ANNEX_USE_GIT_SSH", "1") // withKeyFile works by setting GIT_SSH_COMMAND, but git-annex only respects that if this is set
```

Comment:
This should happen _after_ saving the previous value, right?

Reply:
whoops yep :)